### PR TITLE
new user registration

### PR DESCRIPTION
**Pull request by patmejia**
# run:

```bash
go test -v -cover -race -bench . ./...
```
## Pull Request Updates:
- Added unit tests for user functions in the bechamel-api
- Updated the README with new test instructions 
- Added a quick testing contribution guide
- Streamlined tests and CI with a `runTests` helper function
- Made several changes and updates throughout the codebase, with a net increase of 253 additions and 66 deletions
  - README.md: Various updates to improve clarity and completeness of instructions, addition of instructions for project test running and coverage reporting
  - documentation/TESTING.md: A new document providing a quick example of adding a new function to test
  - go.mod: Updates to the module dependencies
  - internal/user_access.go: Refactoring of the code for better maintainability and readability
- There was no change in the functionality of the program, the changes were primarily related to adding test capabilities, improving documentation, and cleaning up the code.
Screenshot 2023-06-04 at 8 34 52 PM

### Console print: 
```
➜  bechamel-api git:(patmejia/new-user-registration) ✗ go test -v -cover -race -bench . ./...

?       project-ricotta/bechamel-api    [no test files]
=== RUN   TestAuthorizeUser
=== RUN   TestAuthorizeUser/Correct_credentials
=== RUN   TestAuthorizeUser/Empty_credentials
--- PASS: TestAuthorizeUser (0.00s)
    --- PASS: TestAuthorizeUser/Correct_credentials (0.00s)
    --- PASS: TestAuthorizeUser/Empty_credentials (0.00s)
=== RUN   TestGetUserByID
=== RUN   TestGetUserByID/Existing_user_ID
=== RUN   TestGetUserByID/Non-existing_user_ID
--- PASS: TestGetUserByID (0.00s)
    --- PASS: TestGetUserByID/Existing_user_ID (0.00s)
    --- PASS: TestGetUserByID/Non-existing_user_ID (0.00s)
=== RUN   TestGetUserByUserName
=== RUN   TestGetUserByUserName/Existing_username
=== RUN   TestGetUserByUserName/Non-existing_username
--- PASS: TestGetUserByUserName (0.00s)
    --- PASS: TestGetUserByUserName/Existing_username (0.00s)
    --- PASS: TestGetUserByUserName/Non-existing_username (0.00s)
=== RUN   TestAddNewUser
=== RUN   TestAddNewUser/Add_valid_new_user
=== RUN   TestAddNewUser/Add_duplicate_user
--- PASS: TestAddNewUser (0.00s)
    --- PASS: TestAddNewUser/Add_valid_new_user (0.00s)
    --- PASS: TestAddNewUser/Add_duplicate_user (0.00s)
PASS
        project-ricotta/bechamel-api/internal   coverage: 36.5% of statements
ok      project-ricotta/bechamel-api/internal   0.190s
?       project-ricotta/bechamel-api/model      [no test files]
➜  bechamel-api git:(patmejia/new-user-registration) ✗  
```

## Quick Question: 
1. Are we keeping CamelCases for the /model/? 
ie.:
```
➜  model git:(patmejia/new-user-registration) ✗ tree
.
├── LasagnaLoveUser.go
├── LasagnaLoveUserAuthRequest.go
└── LasagnaLoveUser_test.go
```
2. thoughts on database setup/architecture to guide development?

Thank you for your time ⏳

# Roadmap:
-`usersMutex sync.Mutex`: to handle concurrent access to data
-`goconvey`, `testify`, others 🤔
- test for empty, very long inputs, special characters, and more
- [Securing User Data: Concealing the Password Field in JSON Responses](https://github.com/Lasagna-Love-Portal/bechamel-api/issues/19)

### CODE REVIEW

File: documentation/TESTING.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Contributing
```
Comment (mattkuznicki-ll) [inline documentation/TESTING.md]: Hi @patmejia thanks for adding these instructions! Can I suggest this file may be better off in the project root directory rather than in a subfolder? As well, maybe this can form the basis of a DEVELOPMENT.md so we can start consolidating developer instructions from the readme and from testing into one document?

File: README.md [hunk 2]
```diff
@@ -28,74 +28,92 @@ go version
...
+go test -coverprofile=coverage.out ./...
+```
+Coverage Report
+```bash
+go tool cover -html=coverage.out
+```
 ### Contribution
 
-Please see [project-ricotta](https://github.com/Lasagna-Love-Portal/project-ricotta) for more details on contributor guidelines. Contributions are welcomed. Please submit a PR with your changes and they will be reviewed as soon as possible.
+See [project-ricotta](https://github.com/Lasagna-Love-Portal/project-ricotta) for contributor guidelines. Submit a PR for your changes.
```
Comment (mattkuznicki-ll) [inline README.md]: Can you please restore the removed text welcoming contributions?

File: README.md [hunk 3]
```diff
@@ -28,74 +28,92 @@ go version
...
+
+For an examples of how to add new tests, see the [a quick testing guide](./documentation/TESTING.md).
+
 
 ### License
 
-This project is licensed under the terms of the license found in the file [`LICENSE`](LICENSE) in the root directory of this project.
+This project is licensed under the [`LICENSE`](LICENSE) terms.
```
Comment (mattkuznicki-ll) [inline README.md]: Can removed text please be restored to its prior state?

File: README.md [hunk 4]
```diff
@@ -28,74 +28,92 @@ go version
...
 
 ### Contact
 
-For any queries or concerns, please open an issue in the repository or send us an [email](info@lasagnalove.org)
+For queries, please open an issue or [email](mailto:info@lasagnalove.org) us.
```
Comment (mattkuznicki-ll) [inline README.md]: Can altered text please be restored to its prior state?

File: internal/user_access.go [hunk 5]
```diff
@@ -10,59 +10,52 @@ package internal
...
 	{ID: 2, Username: "TestUser2", Password: "password2", GivenName: "Test", FamilyName: "UserTwo"},
 }
 
+func findUser(userFilter func(model.LasagnaLoveUser) bool) (model.LasagnaLoveUser, error) {
+	for _, user := range lasagnaLoveUsers {
+		if userFilter(user) {
+			return user, nil
+		}
+	}
+	return model.LasagnaLoveUser{}, errors.New("no user found with the supplied criteria")
```
Comment (mattkuznicki-ll) [inline internal/user_access.go:26]: Good refactoring here! 👍

File: internal/user_access.go [hunk 6]
```diff
@@ -10,59 +10,52 @@ package internal
...
-	return lasagnaLoveUsers[idx], nil
+
+	return findUser(func(u model.LasagnaLoveUser) bool { return u.Username == userName })
 }
 
 func AddNewUser(newUserProfile model.LasagnaLoveUser) (model.LasagnaLoveUser, error) {
-	// Not allowed to specify an userID - error if one is provided
-	if newUserProfile.ID != 0 {
-		return model.LasagnaLoveUser{}, errors.New("userID may not be specified")
+	if newUserProfile.ID != 0 || newUserProfile.FamilyName == "" || newUserProfile.GivenName == "" ||
```
Comment (mattkuznicki-ll) [inline internal/user_access.go]: This update changes the code from not allowing an ID to be specified, to requiring an ID to be specified. Allowing callers to tell the API what the ID of added records is, is not recommended practice and can introduce security issues.
Can the previous behavior and error for when an ID is provided please be restored?
Also, why the change in error text for an incomplete record? Is this intentional?

File: internal/user_access.go [hunk 7]
```diff
@@ -10,59 +10,52 @@ package internal
...
+		newUserProfile.Username == "" || newUserProfile.Password == "" {
+		return model.LasagnaLoveUser{}, errors.New("invalid or incomplete user data")
 	}
 
-	// Verify required fields are present. Probably an easier way to do this...
-	if newUserProfile.FamilyName == "" || newUserProfile.GivenName == "" ||
-		newUserProfile.Username == "" || newUserProfile.Password == "" {
-		return model.LasagnaLoveUser{}, errors.New("one or more required fields missing or empty")
+	if _, err := GetUserByUserName(newUserProfile.Username); err == nil {
```
Comment (mattkuznicki-ll) [inline internal/user_access.go:61]: Appreciate the check for username existence here! 👍

File: internal/user_access_test.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
+package internal
+
+import (
```
Comment (mattkuznicki-ll) [inline internal/user_access_test.go:3]: Thanks for adding some testing in here! 🥇 

For a future discussion: these tests are likely going to change away from being pure unit tests once there's a database microservice to talk to. We'll want to pay mind to how these tests may evolve if they move away from being pure unit tests to tests that rely on a separate component.

**Review by mattkuznicki-ll — CHANGES_REQUESTED** [review]
Thanks for this PR @patmejia! There's a lot of good stuff in here!

Blocking issue:
* This PR changes the behavior of the profile creation API. Prior behavior did not allow users to specify an ID when making a new entry. This PR seems to change it to requiring users to specify an ID. I submit that we do not want to allow users to specify new IDs for created profiles, the system should be generating these.

Non-blocking questions:
* There are a large number of textual changes in the README that don't seem to really net out to changing anything. Are these intended? I've flagged specific ones that I'd request be undone.
* Would it make it easier for developers to find testing instructions if these were in a top level DEVELOPMENT.md (as we use for developer instructions in the project-ricotta repo)? I think the testing instructions here make a good start to a developer instruction file. We can move some of the developer instructions out of the README and into this in a separate PR to consolidate, then. Moving this file could be a separate PR of course.

File: README.md (hunk 4 above)
Comment (patmejia) [reply README.md]: of course,
please note that the link needs to include the `mailto:` prefix; otherwise, it will not work. I have integrated it accordingly, so it should work no

Comment (patmejia) [reply README.md]: updating to this:

For any queries or concerns, please open an [issue](https://github.com/Lasagna-Love-Portal/bechamel-api/issues) in the repository or send us an email: [info@lasagnalove.org](mailto:info@lasagnalove.org)

File: README.md (hunk 2 above)
Comment (patmejia) [reply README.md]: ✔️ of course

File: README.md (hunk 3 above)
Comment (patmejia) [reply README.md]: ✔️

File: documentation/TESTING.md (hunk 1 above)
Comment (patmejia) [reply documentation/TESTING.md]: - renamed: 
`documentation/TESTING.md`  --> `~/DEVELOPMENT.md`
- sets beginning of our development guide
- ~**pending-to-do**: move guides from README.md to DEVELOPMENT.md~ [reminder/issue#21](https://github.com/Lasagna-Love-Portal/bechamel-api/issues/21)

File: internal/user_access_test.go (hunk 8 above)
Comment (patmejia) [reply internal/user_access_test.go:3]: 💯: Unit -> integration -> functional-> e2e

File: internal/user_access.go (hunk 6 above)
Comment (patmejia) [reply internal/user_access.go]: fixing and updating: insecure direct object references

**Comment by patmejia** [comment]
Screenshot 2023-06-05 at 6 13 56 PM

thank you

**Comment by patmejia** [comment]
I believe I have made all the suggested changes (thanks, @mattkuznicki-ll ). 
Now, open to feedback. 
Please let me know if I missed anything important. 
(Thanks everyone for your efforts)

File: README.md (hunk 4 above)
Comment (mattkuznicki-ll) [reply README.md]: perfect, thank you

File: DEVELOPMENT.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
+# Contributing Guide
```
Comment (mattkuznicki-ll) [inline DEVELOPMENT.md:1]: Thanks for consolidating this into a development file - really helpful! 💯

**Review by mattkuznicki-ll — APPROVED** [review]
Thanks for the updates @patmejia ! Changes look good to me now, will merge
